internal/periodic_checker/publisher.go

```diff
@@ -0,0 +1,384 @@
+func (h TaskHeap) Less(i, j int) bool { return h[i].NextRun.Before(h[j].NextRun) }
+func (h TaskHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
+
+func (h *TaskHeap) Push(x interface{}) {
+	*h = append(*h, x.(*ScheduledTask))
+}
+
+func (h *TaskHeap) Pop() interface{} {
+	old := *h
+	n := len(old)
+	task := old[n-1]
+	*h = old[0 : n-1]
+	return task
+}
+
+func (h *TaskHeap) Peek() *ScheduledTask {
+	if h.Len() == 0 {
+		return nil
+	}
+	return (*h)[0]
+}
+
+func NewTaskHeap() *TaskHeap {
+	h := &TaskHeap{}
+	heap.Init(h)
+	return h
+}
+
+type OrganizationService interface {
+	ListOrganizations(ctx context.Context) (*api.OrganizationList, api.Status)
+}
+
+type TaskChannelManager interface {
+	PublishTask(ctx context.Context, taskRef PeriodicTaskReference) error
+}
+
+type PeriodicTaskPublisher struct {
+	log            logrus.FieldLogger
+	tasksMetadata  map[PeriodicTaskType]PeriodicTaskMetadata
+	orgService     OrganizationService
+	channelManager TaskChannelManager
+	pollConfig     poll.Config
+
+	// Heap for scheduled tasks
+	taskHeap *TaskHeap
+	heapMu   sync.Mutex
+
+	// Organizations tracking
+	organizations map[uuid.UUID]struct{}
+	orgsMu        sync.Mutex
+
+	// Configurable intervals
+	orgSyncInterval time.Duration
+
+	wakeup chan struct{}
+	wg     sync.WaitGroup
+}
+
+type PeriodicTaskPublisherConfig struct {
+	Log             logrus.FieldLogger
+	OrgService      OrganizationService
+	TasksMetadata   map[PeriodicTaskType]PeriodicTaskMetadata
+	ChannelManager  TaskChannelManager
+	OrgSyncInterval time.Duration
+	PollConfig      poll.Config
+}
+
+func NewPeriodicTaskPublisher(publisherConfig PeriodicTaskPublisherConfig) (*PeriodicTaskPublisher, error) {
+	orgSyncInterval := publisherConfig.OrgSyncInterval
+	if orgSyncInterval == 0 {
+		orgSyncInterval = DefaultOrgSyncInterval
+	}
+
+	return &PeriodicTaskPublisher{
+		log:             publisherConfig.Log,
+		orgService:      publisherConfig.OrgService,
+		tasksMetadata:   publisherConfig.TasksMetadata,
+		channelManager:  publisherConfig.ChannelManager,
+		organizations:   make(map[uuid.UUID]struct{}),
+		orgSyncInterval: orgSyncInterval,
+		pollConfig:      publisherConfig.PollConfig,
+		taskHeap:        NewTaskHeap(),
+		wakeup:          make(chan struct{}, 1),
+	}, nil
+}
+
+func (p *PeriodicTaskPublisher) Start(ctx context.Context) {
+	p.log.Info("Starting periodic task publisher")
+	p.wg.Add(1)
+	// Start organization sync goroutine
+	go p.organizationSyncLoop(ctx)
+
+	p.wg.Add(1)
+	// Main scheduling loop
+	go p.schedulingLoop(ctx)
+
+	<-ctx.Done()
+	p.wg.Wait()
+	p.clearHeap()
+	p.log.Info("Periodic task publisher stopped")
+}
+
+func (p *PeriodicTaskPublisher) schedulingLoop(ctx context.Context) {
+	defer p.wg.Done()
+
+	taskTimer := time.NewTimer(DefaultTaskTickerInterval)
+	defer taskTimer.Stop()
+
+	for {
+		p.heapMu.Lock()
+		nextTask := p.taskHeap.Peek()
+		p.heapMu.Unlock()
+
+		if nextTask == nil {
+			taskTimer = time.NewTimer(DefaultTaskTickerInterval)
```

Comment:
Do we need the default ticker interval now that we have a wake up mechanism? Theoretically you either need to sleep until a known event time, or if there are no events, until someone (the org syncing) adds events.